core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt

```diff
@@ -12,382 +12,354 @@ import com.unciv.models.UncivSound
+                        && unit.canUpgrade()) {
+
+                    val goldCostOfUpgrade = unit.getCostOfUpgrade()
+                    val upgradedUnit = unit.getUnitToUpgradeTo()
+
+                    yield(UnitAction(
+                            type = UnitActionType.Upgrade,
+                            title = "Upgrade to [${upgradedUnit.name}] ([$goldCostOfUpgrade] gold)",
+                            uncivSound = UncivSound.Upgrade,
+                            action = getUpgradeAction(unit, tile, upgradedUnit, goldCostOfUpgrade)))
+                }
+
+                // Pillage Action
+                if (!unit.type.isCivilian() && canPillage(unit, tile))
+                    yield(UnitAction(
+                            type = UnitActionType.Pillage,
+                            action = getPillageAction(unit, tile)
+                    ))
+
+                // Setup Action
+                if (!unit.isEmbarked() && unit.hasUnique("Must set up to ranged attack")) {
+                    val isSetUp = unit.action == "Set Up"
+                    yield(UnitAction(
+                            type = UnitActionType.SetUp,
+                            isCurrentAction = isSetUp,
+                            uncivSound = UncivSound.Setup,
+                            action = getLambdaOrNull(unit.currentMovement > 0 && !isSetUp) {
+                                unit.action = Constants.unitActionSetUp
+                                unit.useMovementPoints(1f)
+                            }))
+                }
+
+                // Found city
+                if (!unit.isEmbarked() && unit.hasUnique("Founds a new city"))
+                    yield(UnitAction(
+                            type = UnitActionType.FoundCity,
+                            uncivSound = UncivSound.Chimes,
+                            action = getFoundCityAction(unit, tile)))
+
+                // Worker Actions
+                if (unit.hasUnique("Can build improvements on tiles"))
+                    yieldAll(sequence {
+                        yield(if (Constants.unitActionAutomation == unit.action)
+                            UnitAction(
+                                    type = UnitActionType.StopAutomation,
+                                    action = { unit.action = null })
+                        else
+                            UnitAction(
+                                    type = UnitActionType.Automate,
+                                    action = getLambdaOrNull(unit.currentMovement > 0) {
+                                        unit.action = Constants.unitActionAutomation
+                                        GlobalScope.launch {
+                                            WorkerAutomation(unit).automateWorkerAction()
+                                            UncivGame.Current.worldScreen.shouldUpdate = true
+                                        }
```

Comment:
Here's where the worker automation is on another coroutine so that even if there's a slight delay, it doesn't freeze the UI